### PR TITLE
feat: changed text copy/paste behaviour

### PR DESCRIPTION
**Pull request by adarkforce**
#5783 Changed text copy/paste behaviour by creating one element per line by default and added shortcut to create a single element from multiple lines

## TODOS

- [x] Write tests.
- [x] Improve tests addressing windows/macos differences for copy paste shortcuts.
- [x] Have enough spacing between lines to not make bounding boxes overlap.
- [x] Show plain text paste toast only once per application lifetime.

### CODE REVIEW

**Comment by dwelle** [comment]
We can actually still only handle the `paste` event, we just need to know whether shift was pressed which we can't do from the `ClipboardEvent`. So we need to set a flag in the `keydown` event, which occurs before `paste`. It's kinda hacky, but AFAIK not much else we can do (apart from using the async clipboard API as you did, which isn't ideal).

**Review by adarkforce — COMMENTED** [review]
Yes, it makes sense to have a centralized paste handler and looking for the shiftKey pressed. However setting IS_PLAINTEXT_PASTE to false after 1ms doesn't work on my machine.
What about this at line 1875 of App.tsx?

```
if (
        event[KEYS.CTRL_OR_CMD] &&
        event.key.toLowerCase() === KEYS.V &&
        event.shiftKey
 ) {
        IS_PLAINTEXT_PASTE = true;
} else {
        IS_PLAINTEXT_PASTE = false;
}
```
*edit*
I have realized that this would not be ideal if we could ```paste``` from another source other than key-pressing. That would mean that we should remember every time to set that variabile before calling the pasteFromClipboard function. 
Schermata 2022-10-24 alle 11 58 13

On the other hand the current solution wouldn’t work if we could ``` split-text-paste  ```  from another source other than key-pressing.

Setting the timeout to an higher value won't make us think anymore about that for now, I don't think that there's any human capable of ```Ctrl-Shift-V``` -> ```Ctrl-V```. In less than 100ms I've tested it, it is hard even for mouse clicks :D https://clickspeedtest.com/clicks-per-second.html

**Comment by DanielJGeiger** [comment]
If within the scope of this PR, I would suggest an option for lines to be separate elements, but auto-grouped.

File: src/components/App.tsx [hunk 1]
```diff
@@ -1628,13 +1638,62 @@ class App extends React.Component<AppProps, AppState> {
...
       locked: false,
-    });
+    };
+    const DELTA_GAP = 12;
+    const newElements: ExcalidrawTextElement[] = [];
+    if (splitText) {
+      let currentY = y;
+      const splittedText = text.split("\n");
+      for (const line of splittedText) {
+        if (line.trim().length === 0) {
```
Comment (DanielJGeiger) [inline src/components/App.tsx:1648]: As written, this drops blank lines in pasted text (like for new paragraphs).  Perhaps add a line like `currentY += getApproxLineHeight(font);` before `continue;`?

Comment (adarkforce) [reply src/components/App.tsx:1657]: Yes, corrected. Thank you! :D

**Comment by adarkforce** [comment]
> If within the scope of this PR, I would suggest an option for lines to be separate elements, but auto-grouped.

I need a clarification: do you mean to add a different set of commands for this feature, or to change the current behaviour?

My 2 cents on changing the current behaviour: I bealive this would be a little confusing since the user expectation is to have separate lines. They arleady are selected after pasting so I don’t se the reason to have them grouped on top of that. 

What do @dwelle and @vjeux think about it?

**Comment by vjeux** [comment]
I just played with it and it does exactly what I want. Thanks so much for building this, it's going to help me so much. I used it today for pasting an event schedule I was working on, was nice to be able to get it line by line so that I could move things around.

I'll let @dwelle review the code itself.

**Comment by dwelle** [comment]
Pushed a few changes: 

- remove changelog. We only write changelog for package-related changes. The rest is autogenerated from the commit log.
- reverted inlining of `probablySupportsClipboardReadText`. Tests keep working fine. Otherwise, the course of action would be to try 1) mock tests, or 2) make the `probablySupportsClipboardReadText` into a callable function, instead of inlining, which is error prone if we used that check elsewhere.
- changed how we handle plaintext pasting, so that it doesn't prevent pasting images or elements

I've also changed the line spacing a bit:

- keep spacing consistent regardless of zoom. We don't want to increase/decrease spacing just because you happen to be zoomed in/out.
- add at most one empty line between elements — I've read the comments above, but I think we should not keep multiple empty lines even if in source text. Again, it's about consistency (similar to how e.g. CommonMark handles whitespace) and so that if you copy/paste a spreadsheet or similar data with potentially lots of empty lines, there won't be a huge unnecessary holes. If people show up with counter use cases, we can revert back.
- instead of always using a default line height for spacing, make line space account for current line height. For example, when you paste emoji, the line height is bigger, so the space show be bigger.

Plus some refactor and other small tweaks.

> I have realized that this would not be ideal if we could paste from another source other than key-pressing. That would mean that we should remember every time to set that variabile before calling the pasteFromClipboard function.

We should do both actually. The setTimeout IMO is slightly redundant if we reset on each paste, though as you say it won't cover possible future cases of non-keyboard plaintext-pasting so I kept it.

**Comment by dwelle** [comment]
Alright, sorry for the delay folks. Merging now, tweeting later! 🚀 ❤️

**Review by dwelle — APPROVED** [review]

**Pull request merged**